### PR TITLE
feat: Handle generation numbers in BlobId#{to,from}GsUtilUri

### LINKED ISSUE #1928: Handle generation numbers in BlobId#{to,from}GsUtilUri
Unlike `gsutil`, `BlobId#toGsUtilUri` currently ignores a `BlobId`'s generation number, and `BlobId#fromGsUtilUri` doesn't parse generation numbers. It would be nice if, consistent with `gsutil`, `fromGsUtilUri` parsed any generation number in the URI so that references to particular versions of blobs could be made in URI format.

### PR DESCRIPTION
**Pull request by davidmandle**
Fixes #1928  ☕️

### CODE REVIEW

**Comment by davidmandle** [comment]
@JesseLovelace might you be able to take a look at this? Or is there another code owner I can reach out to?

File: google-cloud-storage/src/main/java/com/google/cloud/storage/BlobId.java
Comment (BenWhitehead) [inline google-cloud-storage/src/main/java/com/google/cloud/storage/BlobId.java:129]: Can we keep each of these argument resolutions on its own line like it was before?

i.e `String bucketName = m.group(1);`

This ensures if we ever get an exception from one of these lines we know which parameter it failed on.

**Review by BenWhitehead — CHANGES_REQUESTED** [review]
In general I think it looks reasonable. One small request on the code structure before approval.

**Comment by davidmandle** [comment]
Thanks for taking a look!

**Comment by davidmandle** [comment]
This appears to have been impacted by a test flake: https://github.com/googleapis/java-storage/issues/1748 .... tests may need to be re-run.

**Comment by davidmandle** [comment]
I believe this is currently blocked on approval of the requested changes.

**Comment by BenWhitehead** [comment]
Change not impacted by owlbot, merging

**Pull request merged**
